Feature/FA-2-Forget-Password

EslamMagdy12 opened the pull request:
This pull request implements the **Forget Password**, **Email Verification**, and **Reset Password** functionalities, integrating them seamlessly across the **data**, **domain**, and **presentation** layers. These enhancements ensure a smoother password recovery experience for users while improving code maintainability and structure. Additionally, minor optimizations and refactoring were done to enhance performance and maintain consistency across the project.

### Changes Included:

#### 🔹 Data Layer:
- Added API endpoints in `AuthRetrofitClient` for:
  - Sending a forget password request.
  - Verifying the reset code.
  - Resetting the password.
- Implemented **Auth Remote Data Source** to manage authentication-related API calls.
- Created request and response models for:
  - `forget_password`
  - `verify_reset_code`
  - `reset_password`
- Developed the **Auth Repository Implementation** to handle data fetching and processing.

#### 🔹 Domain Layer:
- Created new **use cases** for handling:
  - Forget password requests.
  - Reset code verification.
  - Password reset submission.
- Integrated **repository contracts** to ensure structured data handling.

#### 🔹 Presentation Layer:
- Developed dedicated screens for:
  - **Forget Password**
  - **Email Verification**
  - **Reset Password**
- Enhanced UI elements to provide a **more user-friendly experience**.
- Improved widget performance by using `const` where applicable.

#### 🔹 Localization & UI Enhancements:
- Updated Arabic & English translation files (`ar.json` & `en.json`) to include **new authentication-related strings**.
- Introduced **PIN input theme customizations** in `app_theme.dart`.

#### 🔹 General Improvements:
- **Refactored API Exception Handling:**
  - Renamed `app_exception.dart` to `api_exception.dart` for better clarity.
  - Updated all related imports accordingly.
  - Improved `_extractErrorMessage` in `ApiManager` to **handle API errors more effectively**.
- **Navigation Updates:**
  - Registered **Forget Password**, **Email Verification**, and **Reset Password** routes in `routes.dart`.

## 📌 Pull Request Checklist

### ✅ General
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] The PR title follows the required format: `type/FA-<issue-number>-TaskName`.
- [x] The branch name follows the required format: `feature/meaningful-name`, `fix/issue-description`, etc.

### 📂 Code Quality
- [x] All filenames are in `snake_case`.
- [x] All functions and variable names follow Dart naming conventions.
- [x] All new code is formatted with `dart format .`

### 🛠 CI/CD Validation
- [x] Linting passes with `dart analyze`.
- [x] The branch name is validated against the allowed patterns.
- [x] The PR title is validated against the required format.

### 🚀 Testing
- [ ] Unit tests are included for all new features and bug fixes.
- [ ] `flutter test --coverage` has been run and the results reviewed.
- [ ] The project builds successfully without errors.

### 📝 Additional Information
This update enhances the **password recovery process**, ensuring a structured and scalable implementation while improving user experience. Future updates will focus on adding **unit tests** to validate functionality and maintain reliability.

### 🔗 Related JIRA Ticket(s):
- [x] JIRA Issue: `FA-2`

### 📸 Screenshots (if applicable):
<div align="center">

|         Forget Password          |        Email Verification        |        Reset Password         |
|:---------------------:|:----------------------:|:----------------------:|
| <img src="https://github.com/user-attachments/assets/b0d40b43-e4cc-4f02-8a5c-7520fc83532c" width="300" alt="Forget Password"> | <img src="https://github.com/user-attachments/assets/38220ef3-d4ee-49f2-9cad-2aa2a81d43bc" width="300" alt="Email Verification"> | <img src="https://github.com/user-attachments/assets/969ce0bf-cd11-4c60-87a3-f2017cf35a2c" width="300" alt="Reset Password"> |

</div>

EslamMagdy12 committed:
feat(auth): implement forget and reset password data layer.

EslamMagdy12 committed:
feat(auth): implement forget and reset password domain layer.

EslamMagdy12 committed:
Merge branch 'refs/heads/developement' into Feature/FA-2-Forget-Password

EslamMagdy12 committed:
feat(auth): implement forget and reset password presentation layer using cubit as a state management.

EslamMagdy12 committed:
refactor(api): rename app_exception.dart to api_exception.dart and displayed error message returned from api.

EslamMagdy12 committed:
refactor(auth): update widgets to use const for better performance, remove unused imports, and remove relying on hardcoded colors and used the app_colors.

EslamMagdy12 committed:
refactor(auth): reformat and analyze codes using dart format and dart analyze.

EslamMagdy12 committed:
refactor(auth): Refactor Forget Password Feature
- All fields are required in request models.
- Added some missed async and await.
- Removed Pintheme from AppTheme.
- Removed BlocConsumer and used BlocListener Instead.
- Renamed again ApiException to AppException.
- Followed the same Architecture so moved Auth Remote DataSource Impl outside implementation folder

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/reset_password_cubit.dart:
```diff
@@ -0,0 +1,55 @@
+  ResetPasswordCubit(this._resetPasswordUsecase)
+    : super(ResetPasswordState(baseState: BaseInitialState()));
+
+  final TextEditingController passwordController = TextEditingController();
+  final TextEditingController confirmPasswordController =
+      TextEditingController();
```
use dispose function to avoid memory leaks

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/reset_password_cubit.dart:
```diff
@@ -0,0 +1,55 @@
+@injectable
+class ResetPasswordCubit extends Cubit<ResetPasswordState> {
+  final ResetPasswordUsecase _resetPasswordUsecase;
+  final Validator validator = Validator();
```
use constructor injection is better

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/forget_password_cubit.dart:
```diff
@@ -0,0 +1,58 @@
+@injectable
+class ForgetPasswordCubit extends Cubit<ForgetPasswordState> {
+  final ForgetPasswordUsecase forgetPasswordUsecase;
+  final Validator validator = Validator();
```
use constructor injection is better

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/email_verification_state.dart:
```diff
@@ -0,0 +1,28 @@
+import 'package:ecommerce_flower_app/core/base/base_state.dart';
+
+class EmailVerificationState {
+  final BaseState? baseState;
+  final BaseState? resendState;
```
use equtable to improve performance

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/forget_password_state.dart:
```diff
@@ -0,0 +1,15 @@
+import 'package:ecommerce_flower_app/core/base/base_state.dart';
+
+class ForgetPasswordState {
+  final BaseState? baseState;
```
use equtable to improve performance

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/reset_password_state.dart:
```diff
@@ -0,0 +1,19 @@
+import '../../../../core/base/base_state.dart';
+
+class ResetPasswordState {
+  final BaseState? baseState;
```
use equtable to improve performance

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/email_verification_cubit.dart:
```diff
@@ -0,0 +1,82 @@
+    this.forgetPasswordUsecase,
+  ) : super(EmailVerificationState(baseState: BaseInitialState()));
+
+  final TextEditingController pinController = TextEditingController();
```
use dispose function to avoid memory leaks

ahmedelazab1220 commented on lib/features/auth/presentation/view_model/email_verification_cubit.dart:
```diff
@@ -0,0 +1,82 @@
+class EmailVerificationCubit extends Cubit<EmailVerificationState> {
+  final VerifyResetCodeUsecase verifyResetCodeUsecase;
+  final ForgetPasswordUsecase forgetPasswordUsecase;
+  final Validator validator = Validator();
```
use constructor injection is better

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,89 @@
+                          errorText!,
+                          style: TextStyle(
+                            color: AppColors.red,
+                            fontSize: 13.sp,
```
13.sp -> avoid use this not work all screens

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,89 @@
+                },
+              ),
+            ),
+            SizedBox(height: 24.h),
```
this height is fixed and don't need to make responsive because this to lead to inconsistent design

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,89 @@
+                errorPinTheme: AppTheme.errorPinTheme,
+                errorBuilder: (errorText, pin) {
+                  return Padding(
+                    padding: EdgeInsets.only(top: 4.h),
```
this padding is fixed and don't need to make responsive because this to lead to inconsistent design

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,89 @@
+                length: 6,
+                crossAxisAlignment: CrossAxisAlignment.center,
+                onCompleted: (pin) {
+                  context.read<EmailVerificationCubit>().doIntent(
+                    EmailVerificationRequestAction(),
```
this not good to read cubit you can make variable shared on file and access it in file this is better for readability.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,89 @@
+              child: Pinput(
+                keyboardType: TextInputType.number,
+                controller:
+                    context.read<EmailVerificationCubit>().pinController,
+                key: context.read<EmailVerificationCubit>().formKey,
```
this not good to read cubit you can make variable shared on file and access it in file this is better for readability.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,89 @@
+class VerificationCodeInput extends StatelessWidget {
+  const VerificationCodeInput({super.key});
+
+  static PinTheme defaultPinTheme = PinTheme(
+    width: 74.w,
+    height: 68.h,
+    textStyle: TextStyle(fontSize: 24.sp, fontWeight: FontWeight.w500),
+    decoration: BoxDecoration(borderRadius: BorderRadius.circular(10).r),
+  );
+
+  static PinTheme errorPinTheme = defaultPinTheme.copyWith(
+    decoration: BoxDecoration(border: Border.all(color: AppColors.red)),
+  );
```
this is not place for pinTheme you can set in it pin is better

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,79 @@
+                        ),
+                autovalidateMode: AutovalidateMode.onUserInteraction,
+              ),
+              SizedBox(height: 48.h),
```
this is fixed and don't need to make responsive because this to lead to inconsistent design

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,79 @@
+              ),
+              SizedBox(height: 24.h),
+              TextFormField(
+                controller:
+                    context
+                        .read<ResetPasswordCubit>()
+                        .confirmPasswordController,
```
this not good to read cubit you can make variable shared on file and access it in file this is better for readability.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,79 @@
+                  hintText: LocaleKeys.EnterYourNewPassword.tr(),
+                ),
+                validator:
+                    (value) => context
+                        .read<ResetPasswordCubit>()
+                        .validator
+                        .validatePassword(value ?? ''),
```
this not good to read cubit you can make variable shared on file and access it in file this is better for readability.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/reset_password_form.dart:
```diff
@@ -0,0 +1,79 @@
+    return Column(
+      children: [
+        Form(
+          key: context.read<ResetPasswordCubit>().formKey,
```
this not good to read cubit you can make variable shared on file and access it in file this is better for readability.

ahmedelazab1220 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,89 @@
+  const VerificationCodeInput({super.key});
+
+  static PinTheme defaultPinTheme = PinTheme(
+    width: 74.w,
+    height: 68.h,
+    textStyle: TextStyle(fontSize: 24.sp, fontWeight: FontWeight.w500),
```
this is not better way to make it because this to lead to inconsistent design, you can use flexiable is better. 

and textStyle is not suitable make 24.sp is not good way and not work on all screens.

EslamMagdy12 committed:
refactor(auth): refactor forget password

- Used Cubits with BaseState, and Equatable.
- Renamed forget_password_header to header because it used across many screens not just forget_password

EslamMagdy12 committed:
feat(test): Add unit tests for auth use cases

- Added unit tests for ForgetPasswordUsecase, ResetPasswordUsecase, and VerifyResetCodeUsecase.
- Utilized Mockito for mocking the AuthRepo dependency.
- Ensured tests cover both success and failure scenarios for each use case.
- Updated dependencies and configurations to support testing.
- Removed the default widget test file.

EslamMagdy12 committed:
Merge remote-tracking branch 'origin/developement' into feature/fa-2-forget-password

# Conflicts:
#	lib/core/utils/di/di.config.dart
#	lib/core/utils/routes/routes.dart
#	lib/main.dart
#	pubspec.lock
#	test/widget_test.dart

ali72-20 commented on lib/features/auth/data/data_source/remote/auth_remote_data_source_impl.dart:
```diff
@@ -0,0 +1,37 @@
+
+@Injectable(as: AuthRemoteDataSource)
+class AuthRemoteDataSourceImpl implements AuthRemoteDataSource {
+  final AuthRetrofitClient authRetrofitClient;
```
make this private make it better because no one can access authRetroficClient except auth remote data source

ali72-20 commented on lib/features/auth/presentation/view/widgets/resend_code.dart:
```diff
@@ -0,0 +1,83 @@
+                    ),
+                  )
+                  : Text(
+                    'Resend in ${countdown}s',
```
avoid hard coded string

ali72-20 commented on lib/features/auth/presentation/view/widgets/verification_code_input.dart:
```diff
@@ -0,0 +1,94 @@
+                    fontWeight: FontWeight.w500,
+                  ),
+                  decoration: BoxDecoration(
+                    border: Border.all(color: const Color(0xFFCC1010)),
```
try any color move it in app colors to make this color resusble

ali72-20 reviewed (approved):
great job eslam perfect style 
keep going

EslamMagdy12 committed:
refactor(auth): Refactor Forget Password
- Updated forget password screens and widgets to depend on AppColors file
- Update localization files to include (Resend In) text

EslamMagdy12 committed:
refactor(auth): Use private variables in Cubits and Data Sources
- Replaced public variables with private ones in:
  - AuthRemoteDataSourceImpl
  - EmailVerificationCubit
  - ForgetPasswordCubit
- This enhances encapsulation and prevents unintended external access.

EslamMagdy12 committed:
test(auth): Implement Forget Password Feature with Unit Tests
- Included unit tests for the following components:
  - Forget Password Cubit
  - Auth Repository Implementation
  - Auth Remote Data Source Implementation
  - Forget Password Usecase
  - Reset Password Usecase
  - Verify Reset Code Usecase

EslamMagdy12 committed:
refactor(test): deleted anything related to testing

EslamMagdy12 committed:
refactor(auth): dispose timer in ResendCode widget
- Prevents memory leaks by canceling the timer in the `dispose` method.

EslamMagdy12 committed:
Merge remote-tracking branch 'origin/developement' into feature/fa-2-forget-password

# Conflicts:
#	assets/translations/ar.json
#	assets/translations/en.json
#	lib/core/utils/datasource_excution/api_manager.dart
#	lib/core/utils/di/di.config.dart
#	lib/core/utils/routes/routes.dart
#	lib/features/auth/data/data_source/contract/auth_remote_data_source.dart
#	lib/features/auth/data/data_source/remote/auth_remote_data_source_impl.dart
#	lib/features/auth/data/repo_impl/auth_repo_impl.dart
#	lib/features/auth/domain/repo/auth_repo.dart
#	lib/features/home/data/api/home_retrofit_client.dart
#	lib/features/home/data/api/home_retrofit_client.g.dart
#	lib/main.dart
#	pubspec.lock
#	pubspec.yaml

EslamMagdy12 committed:
refactor(auth): merged dev into forget password branch, and fixed some issues

EslamMagdy12 committed:
refactor(auth): reformated using dart format --set-exit-if-changed . and ensured no issues when did dart analyze

EslamMagdy12 committed:
refactor(auth): standardize use case naming and improve import paths

ahmedelazab1220 committed:
Merge pull request #60 from ahmedelazab1220/developement

Feature/FA-100-Merge-Final-One

EslamMagdy12 committed:
Merge remote-tracking branch 'refs/remotes/origin/developement' into feature/fa-2-forget-password

# Conflicts:
#	assets/translations/ar.json
#	assets/translations/en.json
#	lib/core/utils/di/di.config.dart
#	lib/core/utils/routes/routes.dart
#	pubspec.yaml

EslamMagdy12 committed:
fix(routes): update initial route to use dynamic route initializer.

EslamMagdy12 committed:
refactor(forget-password): merged dev to fix conflicts and ensured dart analysis no issues

EslamMagdy12 committed:
refactor(forget-password): update forget password flow and improve UI components

EslamMagdy12 committed:
refactor(forget-password): simplify button structure in forget and reset password forms

EslamMagdy12 committed:
fix(theme, dependencies): update dialogTheme to use DialogThemeData for consistency, and updated skeletonizer package.

ahmedelazab1220 committed:
Create LICENSE

ahmedelazab1220 committed:
docs(README): enhance project overview and architecture details
chore(pubspec): update package versions in pubspec.lock

ahmedelazab1220 committed:
Merge branch 'main' of github.com:ahmedelazab1220/ecommerce_flower_app

ahmedelazab1220 committed:
docs(README): add additional sources for dependency injection configuration

ahmedelazab1220 committed:
Update README.md

Fix link of demo video

ahmedelazab1220 committed:
Update README.md

Commit:
Merge branch 'main' into developement

Commit:
Merge branch 'main' into feature/fa-2-forget-password

Commit:
Merge branch 'developement' into feature/fa-2-forget-password
